addons/misra.py

```diff
@@ -256,14 +256,105 @@ def getForLoopExpressions(forToken):
             lpar.astOperand2.astOperand2.astOperand2]
 
 
+def findCounterTokens(cond):
+    if cond.str in ['&&', '||']:
+        c = findCounterTokens(cond.astOperand1)
+        c.extend(findCounterTokens(cond.astOperand2))
+        return c
+    ret = []
+    if cond.isArithmeticalOp:
+        if cond.astOperand1.isName and cond.astOperand2.isNumber:
+            ret.append(cond.astOperand1)
+        elif cond.astOperand2.isName and cond.astOperand1.isNumber:
+            ret.append(cond.astOperand2)
+        elif cond.astOperand2.isName and cond.astOperand1.isName:
+            ret.extend([cond.astOperand2, cond.astOperand1])
+        elif cond.astOperand1.isOp and cond.astOperand2.isName:
+            ret.extend(findCounterTokens(cond.astOperand1))
+            ret.append(cond.astOperand2)
+        elif cond.astOperand2.isOp and cond.astOperand1.isName:
+            ret.extend(findCounterTokens(cond.astOperand2))
+            ret.append(cond.astOperand1)
+        elif cond.astOperand2.isOp and cond.astOperand1.isOp:
+            ret.extend(findCounterTokens(cond.astOperand1))
+            ret.extend(findCounterTokens(cond.astOperand2))
+        elif cond.astOperand1.isOp and cond.astOperand2.isNumber:
+            ret.extend(findCounterTokens(cond.astOperand1))
+        elif cond.astOperand2.isOp and cond.astOperand1.isNumber:
+            ret.extend(findCounterTokens(cond.astOperand2))
+
+    if cond.isComparisonOp:
+        if cond.astOperand1.isName and cond.astOperand2.isNumber:
+            ret.append(cond.astOperand1)
+        elif cond.astOperand2.isName and cond.astOperand1.isNumber:
+            ret.append(cond.astOperand2)
+        elif cond.astOperand2.isOp and cond.astOperand1.isOp:
+            ret.extend(findCounterTokens(cond.astOperand1))
+            ret.extend(findCounterTokens(cond.astOperand2))
+        elif cond.astOperand1.isOp and cond.astOperand2.isNumber:
+            ret.extend(findCounterTokens(cond.astOperand1))
+        elif cond.astOperand2.isOp and cond.astOperand1.isNumber:
+            ret.extend(findCounterTokens(cond.astOperand2))
+
+    return ret
+
+
+def getFloatComparision(floatCompOperator):
+    while not floatCompOperator.isComparisonOp:
+        floatCompOperator = floatCompOperator.astParent
+    expr = hasFloatComparison(floatCompOperator)
+    if expr:
+        return True
+    return False
+
+
+def getWhileLoopExpressions(whileToken):
+    if not simpleMatch(whileToken, 'while ('):
+        return None
+    lpar = whileToken.next
+    if not lpar.astOperand2.astOperand1:
+        return None
+    rpar = lpar.link
+    counter = findCounterTokens(lpar.astOperand2)
+    if len(counter) == 0:
+        return None
+    if simpleMatch(rpar, ') {'):
+        token = rpar.next
+        while (token != rpar.next.link):
```

Comment:
this loop is more or less copy pasted and written twice. please refactor it so there is only 1 loop that looks at the while body.

Reply:
I still think this is true.

I envision something like:

    whileBodyStart = None
    if simpleMatch(rpar, ') {'):
        whileBodyStart = rpar.next
    elif simpleMatch(whileToken.previous, '} while') and simpleMatch(whileToken.previous.link.previous, 'do {'):
        whileBodyStart = whileToken.previous.link
    else:
        return False

    ... check the while loop body ...

Reply:
Wow!!
I never thought this way!
Thanks a lot  :)